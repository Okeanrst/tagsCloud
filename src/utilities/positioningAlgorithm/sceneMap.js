import IntersectionError from './IntersectionError';

const X = 'x';
const MINUS_X = '-x';
const Y = 'y';
const MINUS_Y = '-y';

function sceneMap() {
	this.sceneMap = {
		'xy': [],
		'x-y': [],
		'-x-y': [],
		'-xy': []
	};
	this.sizeX = 0;
	this.sizeMinusX = 0;
	this.sizeY = 0;
	this.sizeMinusY = 0;	
}

function calcQuarter(xIsPositive, yIsPositive) {
	return `${xIsPositive ? '' : '-'}x${yIsPositive ? '' : '-'}y`;
}

sceneMap.prototype.getSceneSize = function getSceneSize() {
	const {sizeX, sizeMinusX, sizeY, sizeMinusY} = this;	
	return {[X]: sizeX, [MINUS_X]: sizeMinusX, [Y]: sizeY, [MINUS_Y]: sizeMinusY};
}

sceneMap.prototype.calcSceneSize = function getSceneSize() {
  //Y
	if (this.sceneMap['xy'].length > this.sizeY) {
    this.sizeY = this.sceneMap['xy'].length;
	}

  if (this.sceneMap['-xy'].length > this.sizeY) {
    this.sizeY = this.sceneMap['-xy'].length;
  }
	//-Y
  if (this.sceneMap['x-y'].length > this.sizeMinusY) {
    this.sizeMinusY = this.sceneMap['x-y'].length
  }
  if (this.sceneMap['x-y'].length > this.sizeMinusY) {
    this.sizeMinusY = this.sceneMap['x-y'].length
  }

  //X
  this.sceneMap['xy'].forEach(row => {
    if (row.length > this.sizeX) {
      this.sizeX = row.length;
    }
  });
  this.sceneMap['x-y'].forEach(row => {
    if (row.length > this.sizeX) {
      this.sizeX = row.length;
    }
  });
	//-X
  this.sceneMap['-xy'].forEach(row => {
    if (row.length > this.sizeMinusX) {
      this.sizeMinusX = row.length;
    }
  });
  this.sceneMap['-x-y'].forEach(row => {
    if (row.length > this.sizeMinusX) {
      this.sizeMinusX = row.length;
    }
  });
}

sceneMap.prototype._setDataAtPosition = function setDataAtPosition(x, y, val = true) {
  const xIsPositive = x >= 0;
  const yIsPositive = y >= 0;
  const quarter = calcQuarter(xIsPositive, yIsPositive);
  const row = Math.abs(y);
  const col = Math.abs(x);
  if (!Array.isArray(this.sceneMap[quarter][Math.abs(y)])) {
    this.sceneMap[quarter][row] = [];
  }
  this.sceneMap[quarter][row][col] = val;
}

sceneMap.prototype.setDataAtPosition = function setDataAtPosition(x, y, val = true) {
	if (typeof x !== 'number' || typeof y !== 'number') {
		throw new Error('setDataAtPosition error: typeof x !== number || typeof y !== number');
	}
	if (x === 0 || y === 0) {
		throw new Error('setDataAtPosition error: x === 0 || y === 0');
	}
	if (this.getDataAtPosition(x, y)) {
		throw new IntersectionError();
	}
  this._setDataAtPosition(x, y);
}

sceneMap.prototype.releasePosition = function(x, y) {
  this._setDataAtPosition(x, y, undefined);
}

sceneMap.prototype.getDataAtPosition = function getDataAtPosition(x, y) {
	if (typeof x !== 'number' || typeof y !== 'number') {
		throw new Error('getDataAtPosition error: typeof x !== number || typeof y !== number');
	}
	if (x === 0 || y === 0) {
		throw new Error('getDataAtPosition error: x === 0 || y === 0');
	}	
	const quarter = calcQuarter(x >= 0, y >= 0);
	const row = Math.abs(y);
	const col = Math.abs(x);
	if (!Array.isArray(this.sceneMap[quarter][row])) {
		return;
	}	
	return this.sceneMap[quarter][row][col];
}

sceneMap.prototype.drawItself = function drawItself() {
	//сверху вниз
	const sceneSize = this.getSceneSize();
	const topRow = sceneSize[Y];
	const bottomRow = -sceneSize[MINUS_Y];
	const leftCol = -sceneSize[MINUS_X];
	const rightCol = sceneSize[X];
	
	let res = '';	

	for (let row = topRow; row >= bottomRow; row--) {				
		for (let col = leftCol; col <= rightCol; col++) {
			if (row === 0) {
				res +=  '-';				
			} else {
				res += col === 0 ? '|' : this.getDataAtPosition(col, row) ? '#' : '.';
			}
		}
		res += '\n';
	}

	console.log(res, '\n\n');	
}

sceneMap.rectSizeToPositionUnits = function (rectSize, ratio) {
	return Math.ceil(rectSize / ratio);
}

sceneMap.rectToSceneMap = function rectToSceneMap(rectVal, ratio) {
	return Math.sign(rectVal)*Math.ceil(Math.abs(rectVal / ratio));
}

sceneMap.sceneMapToRect = function sceneMapToRect(mapVal, ratio) {
	return mapVal * ratio;
}

sceneMap.countPositions = function(begin, end) {
	if (begin > end) {
    throw new Error('countPositions error: can not be begin > end');
	}
  if (begin === 0 || end === 0) {
    throw new Error('countPositions error: begin, end can not be zero');
  }
	return begin < 0 && end > 0 ? end - begin : end - begin + 1;
}

sceneMap.takePositionsFromFirst = function(first, count) {
  if (count <= 0) throw new Error('"count" must be positive');
  if (first === 0) throw new Error('"first" can not be zero');
  if (count === 1) return first;
  let res = first + count;
  if (!(first < 0 && res >= 0)) {
    res--;
  }
  return res;
}

sceneMap.takePositionsFromLast = function(last, count) {
  if (count <= 0) throw new Error('"count" must be positive');
  if (last === 0) throw new Error('"last" can not be zero');
  if (count === 1) return last;

  let res = last - count;
  if (res >= 0 || last < 0) {
    res++;
  }
  return res;
}

sceneMap.nextPosition = cur => cur === -1 ? cur + 2 : cur + 1;
sceneMap.prevPosition = cur => cur === 1 ? cur - 2 : cur - 1;
sceneMap.changePosition = (cur, diff) => {
  if (cur === 0) throw new Error('"cur" can not be zero');
  let res = cur + diff;
  if (cur > 0 && res <= 0) {
    res--;
  }
  if (cur < 0 && res >= 0) {
    res++;
  }
  return res;
}

export default sceneMap;
export const dimensions = {X, MINUS_X, MINUS_Y, Y};