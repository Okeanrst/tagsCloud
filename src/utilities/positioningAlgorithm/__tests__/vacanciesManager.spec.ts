import { VacanciesManager } from '../vacanciesManager';
import { SceneMap } from '../sceneMap';

describe('VacanciesManager tests', () => {
  let sceneMap: SceneMap;
  let vacanciesManager: VacanciesManager;
  beforeEach(() => {
    sceneMap = new SceneMap();
    vacanciesManager = new VacanciesManager(sceneMap);
  });
  it('correct vacancies for empty sceneMap', () => {
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    expect(vacanciesManager.closedVacancies).toStrictEqual([]);
    expect(vacanciesManager.topEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.rightEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.bottomEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.leftEdgeVacancies).toStrictEqual([]);
  });
  it('correct vacancies for one rect laid at sceneMap', () => {
    sceneMap.bulkUpdate([
      [1, 1, true],
      [2, 1, true],
      [1, 2, true],
      [2, 2, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    expect(vacanciesManager.closedVacancies).toStrictEqual([]);
    expect(vacanciesManager.topEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.rightEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.bottomEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.leftEdgeVacancies).toStrictEqual([]);
  });
  it('correct vacancies for one rect laid at sceneMap', () => {
    sceneMap.bulkUpdate([
      [1, 1, true], [2, 1, true], [3, 1, true],
      [1, 2, true], [3, 2, true],
      [1, 3, true], [2, 3, true], [3, 3, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    expect(vacanciesManager.closedVacancies).toStrictEqual([
      {
        bottom: 2,
        cols: 1,
        distanceFromCenter: 0.7071067811865476,
        left: 2,
        right: 2,
        rows: 1,
        square: 1,
        top: 2,
      },
    ]);
    expect(vacanciesManager.topEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.rightEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.bottomEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.leftEdgeVacancies).toStrictEqual([]);
  });
  it('correct vacancies in the top right conner of the sceneMap', () => {
    sceneMap.bulkUpdate([
      [1, 1, true],
      [2, 1, true],
      [1, 2, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const topRightEdgeVacancy = {
      baseSize: Infinity,
      bottom: 2,
      distanceFromCenter: 1.4142135623730951,
      left: 2,
      right: Infinity,
      rightEdgeColumn: 2,
      top: Infinity,
      topEdgeRow: 2,
    };
    expect(vacanciesManager.closedVacancies).toStrictEqual([]);
    expect(vacanciesManager.topEdgeVacancies).toStrictEqual([
      topRightEdgeVacancy,
    ]);
    expect(vacanciesManager.rightEdgeVacancies).toStrictEqual([
      topRightEdgeVacancy,
    ]);
    expect(vacanciesManager.bottomEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.leftEdgeVacancies).toStrictEqual([]);
  });
  it('correct vacancies in the bottom right conner of the sceneMap', () => {
    sceneMap.bulkUpdate([
      [1, -1, true],
      [2, -1, true],
      [1, -2, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const bottomRightEdgeVacancy = {
      baseSize: Infinity,
      bottom: -Infinity,
      bottomEdgeRow: -2,
      distanceFromCenter: 1.4142135623730951,
      left: 2,
      right: Infinity,
      rightEdgeColumn: 2,
      top: -2,
    };
    expect(vacanciesManager.closedVacancies).toStrictEqual([]);
    expect(vacanciesManager.topEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.rightEdgeVacancies).toStrictEqual([
      bottomRightEdgeVacancy,
    ]);
    expect(vacanciesManager.bottomEdgeVacancies).toStrictEqual([
      bottomRightEdgeVacancy,
    ]);
    expect(vacanciesManager.leftEdgeVacancies).toStrictEqual([]);
  });
  it('correct vacancies in the bottom left conner of the sceneMap', () => {
    sceneMap.bulkUpdate([
      [-1, -1, true],
      [-2, -1, true],
      [-1, -2, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const bottomLeftEdgeVacancy = {
      baseSize: Infinity,
      bottom: -Infinity,
      bottomEdgeRow: -2,
      left: -Infinity,
      leftEdgeColumn: -2,
      right: -2,
      top: -2,
      distanceFromCenter: 1.4142135623730951,
    };
    expect(vacanciesManager.closedVacancies).toStrictEqual([]);
    expect(vacanciesManager.topEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.rightEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.bottomEdgeVacancies).toStrictEqual([
      bottomLeftEdgeVacancy,
    ]);
    expect(vacanciesManager.leftEdgeVacancies).toStrictEqual([
      bottomLeftEdgeVacancy,
    ]);
  });
  it('correct vacancies in the top left conner of the sceneMap', () => {
    sceneMap.bulkUpdate([
      [-1, 1, true],
      [-2, 1, true],
      [-1, 2, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const topLeftEdgeVacancy = {
      baseSize: Infinity,
      bottom: 2,
      left: -Infinity,
      leftEdgeColumn: -2,
      right: -2,
      top: Infinity,
      topEdgeRow: 2,
      distanceFromCenter: 1.4142135623730951,
    };
    expect(vacanciesManager.closedVacancies).toStrictEqual([]);
    expect(vacanciesManager.topEdgeVacancies).toStrictEqual([
      topLeftEdgeVacancy,
    ]);
    expect(vacanciesManager.rightEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.bottomEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.leftEdgeVacancies).toStrictEqual([
      topLeftEdgeVacancy,
    ]);
  });
  it('correct vacancy at 1, 1 position of the sceneMap', () => {
    sceneMap.bulkUpdate([
      [2, 1, true],
      [1, 2, true],
      [2, 2, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const bottomLeftEdgeVacancy = {
      baseSize: Infinity,
      bottom: -Infinity,
      bottomEdgeRow: 1,
      left: -Infinity,
      leftEdgeColumn: 1,
      right: 1,
      top: 1,
      distanceFromCenter: 0,
    };
    expect(vacanciesManager.closedVacancies).toStrictEqual([]);
    expect(vacanciesManager.topEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.rightEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.bottomEdgeVacancies).toStrictEqual([
      bottomLeftEdgeVacancy,
    ]);
    expect(vacanciesManager.leftEdgeVacancies).toStrictEqual([
      bottomLeftEdgeVacancy,
    ]);
  });
  it('two closed vacancies, test removeClosedVacancy', () => {
    sceneMap.bulkUpdate([
      [-1, -1, true], [1, -1, true], [2, -1, true], [3, -1, true], [4, -1, true],
      [-1, 1, true], [4, 1, true],
      [-1, 2, true], [3, 2, true], [4, 2, true],
      [-1, 3, true], [3, 3, true], [4, 3, true],
      [-1, 4, true], [1, 4, true], [2, 4, true], [3, 4, true], [4, 4, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const closedVacancies = [
      {
        bottom: 1,
        cols: 2,
        distanceFromCenter: 0.5,
        left: 1,
        right: 2,
        rows: 3,
        square: 6,
        top: 3,
      }
    ];

    expect(vacanciesManager.closedVacancies).toMatchInlineSnapshot(`
      Array [
        Object {
          "bottom": 1,
          "cols": 2,
          "distanceFromCenter": 0.5,
          "left": 1,
          "right": 2,
          "rows": 3,
          "square": 6,
          "top": 3,
        },
        Object {
          "bottom": 1,
          "cols": 3,
          "distanceFromCenter": 0.7071067811865476,
          "left": 1,
          "right": 3,
          "rows": 1,
          "square": 3,
          "top": 1,
        },
      ]
    `);
    expect(vacanciesManager.topEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.rightEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.bottomEdgeVacancies).toStrictEqual([]);
    expect(vacanciesManager.leftEdgeVacancies).toStrictEqual([]);
    vacanciesManager.removeClosedVacancy(1);
    expect(vacanciesManager.closedVacancies).toStrictEqual([
      closedVacancies[0],
      undefined,
    ]);
    expect(() => {
      vacanciesManager.removeClosedVacancy(1);
    }).toThrowError(/^vacancy index [\d]+ does not exist$/);
  });
  it('should return correct vacancies for sceneMap 1', () => {
    sceneMap.bulkUpdate([
      [1, 1, true], [2, 1, true], [3, 1, true], [4, 1, true], [5, 1, true], [6, 1, true],
      [1, 2, true], [2, 2, true], [3, 2, true], [6, 2, true],
      [1, 3, true], [2, 3, true], [6, 3, true],
      [1, 4, true], [6, 4, true],
      [1, 5, true], [2, 5, true], [3, 5, true], [4, 5, true], [6, 5, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const {
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    } = vacanciesManager;
    expect({
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    }).toMatchSnapshot();
  });
  it('should return correct vacancies for sceneMap 2', () => {
    sceneMap.bulkUpdate([
      [1, 1, true], [2, 1, true], [3, 1, true], [4, 1, true], [5, 1, true], [6, 1, true],
      [1, 2, true], [2, 2, true], [3, 2, true], [6, 2, true],
      [1, 3, true], [2, 3, true], [6, 3, true],
      [1, 4, true], [6, 4, true],
      [1, 5, true], [2, 5, true], [3, 5, true], [6, 5, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const {
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    } = vacanciesManager;
    expect({
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    }).toMatchSnapshot();
  });
  it('should return correct vacancies for sceneMap 3', () => {
    sceneMap.bulkUpdate([
      [1, 1, true], [2, 1, true], [3, 1, true], [4, 1, true], [5, 1, true], [6, 1, true],
      [1, 2, true], [2, 2, true], [3, 2, true], [4, 2, true], [6, 2, true],
      [1, 3, true], [2, 3, true], [6, 3, true],
      [1, 4, true], [6, 4, true],
      [1, 5, true], [2, 5, true], [3, 5, true], [5, 5, true], [6, 5, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const {
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    } = vacanciesManager;
    expect({
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    }).toMatchSnapshot();
  });
  it('should return correct vacancies for sceneMap 4', () => {
    sceneMap.bulkUpdate([
      [1, -1, true], [6, -1, true],
      [1, 1, true], [2, 1, true], [3, 1, true], [4, 1, true], [5, 1, true], [6, 1, true],
      [1, 2, true], [2, 2, true], [3, 2, true], [4, 2, true], [6, 2, true],
      [1, 3, true], [2, 3, true], [6, 3, true],
      [1, 4, true], [6, 4, true],
      [1, 5, true], [2, 5, true], [3, 5, true], [5, 5, true], [6, 5, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const {
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    } = vacanciesManager;
    expect({
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    }).toMatchSnapshot();
  });
  it('should return correct vacancies for sceneMap 5', () => {
    sceneMap.bulkUpdate([
      [-1, 1, true], [1, 1, true], [2, 1, true], [3, 1, true], [4, 1, true], [5, 1, true],
      [-1, 2, true], [5, 2, true],
      [-1, 3, true], [2, 3, true],
      [-1, 4, true], [2, 4, true], [3, 4, true], [5, 4, true],
      [-1, 5, true], [1, 5, true], [2, 5, true], [3, 5, true], [4, 5, true], [5, 5, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const {
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    } = vacanciesManager;
    expect({
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    }).toMatchSnapshot();
  });
  it('should return correct vacancies for sceneMap 6', () => {
    sceneMap.bulkUpdate([
      [-2, 2, true], [-1, 2, true], [1, 2, true], [2, 2, true], [3, 2, true], [4, 2, true],
      [-2, 1, true], [4, 1, true],
      [-2, -1, true], [-1, -1, true], [4, -1, true],
      [-2, -2, true], [3, -2, true], [4, -2, true],
      [-2, -3, true], [4, -3, true],
    ]);
    sceneMap.calcSceneEdges();
    vacanciesManager.buildVacanciesMap();
    const {
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    } = vacanciesManager;
    expect({
      closedVacancies,
      topEdgeVacancies,
      rightEdgeVacancies,
      bottomEdgeVacancies,
      leftEdgeVacancies
    }).toMatchSnapshot();
  });
});
