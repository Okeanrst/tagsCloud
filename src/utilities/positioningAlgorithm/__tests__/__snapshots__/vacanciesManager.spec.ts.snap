// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VacanciesManager tests should return correct vacancies for sceneMap 1 1`] = `
Object {
  "bottomEdgeVacancies": Array [],
  "closedVacancies": Array [
    Object {
      "bottom": 3,
      "cols": 3,
      "distanceFromCenter": 1.4142135623730951,
      "left": 3,
      "right": 5,
      "rows": 2,
      "square": 6,
      "top": 4,
    },
    Object {
      "bottom": 4,
      "cols": 4,
      "distanceFromCenter": 1.5811388300841898,
      "left": 2,
      "right": 5,
      "rows": 1,
      "square": 4,
      "top": 4,
    },
    Object {
      "bottom": 2,
      "cols": 2,
      "distanceFromCenter": 1.5811388300841898,
      "left": 4,
      "right": 5,
      "rows": 3,
      "square": 6,
      "top": 4,
    },
  ],
  "leftEdgeVacancies": Array [],
  "rightEdgeVacancies": Array [],
  "topEdgeVacancies": Array [
    Object {
      "baseSize": 1,
      "bottom": 2,
      "distanceFromCenter": 2.0615528128088303,
      "left": 5,
      "right": 5,
      "top": Infinity,
      "topEdgeRow": 5,
    },
  ],
}
`;

exports[`VacanciesManager tests should return correct vacancies for sceneMap 2 1`] = `
Object {
  "bottomEdgeVacancies": Array [],
  "closedVacancies": Array [
    Object {
      "bottom": 3,
      "cols": 3,
      "distanceFromCenter": 1.4142135623730951,
      "left": 3,
      "right": 5,
      "rows": 2,
      "square": 6,
      "top": 4,
    },
    Object {
      "bottom": 4,
      "cols": 4,
      "distanceFromCenter": 1.5811388300841898,
      "left": 2,
      "right": 5,
      "rows": 1,
      "square": 4,
      "top": 4,
    },
  ],
  "leftEdgeVacancies": Array [],
  "rightEdgeVacancies": Array [],
  "topEdgeVacancies": Array [
    Object {
      "baseSize": 2,
      "bottom": 2,
      "distanceFromCenter": 1.5811388300841898,
      "left": 4,
      "right": 5,
      "top": Infinity,
      "topEdgeRow": 5,
    },
  ],
}
`;

exports[`VacanciesManager tests should return correct vacancies for sceneMap 3 1`] = `
Object {
  "bottomEdgeVacancies": Array [],
  "closedVacancies": Array [
    Object {
      "bottom": 3,
      "cols": 3,
      "distanceFromCenter": 1.4142135623730951,
      "left": 3,
      "right": 5,
      "rows": 2,
      "square": 6,
      "top": 4,
    },
    Object {
      "bottom": 4,
      "cols": 4,
      "distanceFromCenter": 1.5811388300841898,
      "left": 2,
      "right": 5,
      "rows": 1,
      "square": 4,
      "top": 4,
    },
    Object {
      "bottom": 2,
      "cols": 1,
      "distanceFromCenter": 2.0615528128088303,
      "left": 5,
      "right": 5,
      "rows": 3,
      "square": 3,
      "top": 4,
    },
  ],
  "leftEdgeVacancies": Array [],
  "rightEdgeVacancies": Array [],
  "topEdgeVacancies": Array [
    Object {
      "baseSize": 1,
      "bottom": 3,
      "distanceFromCenter": 1.118033988749895,
      "left": 4,
      "right": 4,
      "top": Infinity,
      "topEdgeRow": 5,
    },
  ],
}
`;

exports[`VacanciesManager tests should return correct vacancies for sceneMap 4 1`] = `
Object {
  "bottomEdgeVacancies": Array [
    Object {
      "baseSize": 4,
      "bottom": -Infinity,
      "bottomEdgeRow": -1,
      "distanceFromCenter": 3.0413812651491097,
      "left": 2,
      "right": 5,
      "top": -1,
    },
  ],
  "closedVacancies": Array [
    Object {
      "bottom": 3,
      "cols": 3,
      "distanceFromCenter": 1.8027756377319946,
      "left": 3,
      "right": 5,
      "rows": 2,
      "square": 6,
      "top": 4,
    },
    Object {
      "bottom": 4,
      "cols": 4,
      "distanceFromCenter": 2.0615528128088303,
      "left": 2,
      "right": 5,
      "rows": 1,
      "square": 4,
      "top": 4,
    },
    Object {
      "bottom": 2,
      "cols": 1,
      "distanceFromCenter": 2.23606797749979,
      "left": 5,
      "right": 5,
      "rows": 3,
      "square": 3,
      "top": 4,
    },
  ],
  "leftEdgeVacancies": Array [],
  "rightEdgeVacancies": Array [],
  "topEdgeVacancies": Array [
    Object {
      "baseSize": 1,
      "bottom": 3,
      "distanceFromCenter": 1.4142135623730951,
      "left": 4,
      "right": 4,
      "top": Infinity,
      "topEdgeRow": 5,
    },
  ],
}
`;

exports[`VacanciesManager tests should return correct vacancies for sceneMap 5 1`] = `
Object {
  "bottomEdgeVacancies": Array [],
  "closedVacancies": Array [
    Object {
      "bottom": 2,
      "cols": 4,
      "distanceFromCenter": 0.7071067811865476,
      "left": 1,
      "right": 4,
      "rows": 1,
      "square": 4,
      "top": 2,
    },
    Object {
      "bottom": 2,
      "cols": 1,
      "distanceFromCenter": 1.118033988749895,
      "left": 1,
      "right": 1,
      "rows": 3,
      "square": 3,
      "top": 4,
    },
    Object {
      "bottom": 2,
      "cols": 2,
      "distanceFromCenter": 1.5,
      "left": 3,
      "right": 4,
      "rows": 2,
      "square": 4,
      "top": 3,
    },
    Object {
      "bottom": 2,
      "cols": 1,
      "distanceFromCenter": 2.0615528128088303,
      "left": 4,
      "right": 4,
      "rows": 3,
      "square": 3,
      "top": 4,
    },
  ],
  "leftEdgeVacancies": Array [],
  "rightEdgeVacancies": Array [
    Object {
      "baseSize": Infinity,
      "bottom": 3,
      "distanceFromCenter": 1.118033988749895,
      "left": 3,
      "right": Infinity,
      "rightEdgeColumn": 5,
      "top": 3,
    },
  ],
  "topEdgeVacancies": Array [],
}
`;

exports[`VacanciesManager tests should return correct vacancies for sceneMap 6 1`] = `
Object {
  "bottomEdgeVacancies": Array [
    Object {
      "baseSize": 3,
      "bottom": -Infinity,
      "bottomEdgeRow": -3,
      "distanceFromCenter": 1.5811388300841898,
      "left": -1,
      "right": 2,
      "top": -2,
    },
    Object {
      "baseSize": 2,
      "bottom": -Infinity,
      "bottomEdgeRow": -3,
      "distanceFromCenter": 1.5811388300841898,
      "left": 1,
      "right": 2,
      "top": 1,
    },
    Object {
      "baseSize": 4,
      "bottom": -Infinity,
      "bottomEdgeRow": -3,
      "distanceFromCenter": 2.5,
      "left": -1,
      "right": 3,
      "top": -3,
    },
  ],
  "closedVacancies": Array [
    Object {
      "bottom": -1,
      "cols": 3,
      "distanceFromCenter": 1.118033988749895,
      "left": 1,
      "right": 3,
      "rows": 2,
      "square": 6,
      "top": 1,
    },
    Object {
      "bottom": 1,
      "cols": 4,
      "distanceFromCenter": 1.5,
      "left": -1,
      "right": 3,
      "rows": 1,
      "square": 4,
      "top": 1,
    },
  ],
  "leftEdgeVacancies": Array [],
  "rightEdgeVacancies": Array [],
  "topEdgeVacancies": Array [],
}
`;

exports[`VacanciesManager tests should return correct vacancies for sceneMap 7 1`] = `
Object {
  "bottomEdgeVacancies": Array [
    Object {
      "baseSize": 1,
      "bottom": -Infinity,
      "bottomEdgeRow": -4,
      "distanceFromCenter": 0.5,
      "left": 4,
      "right": 4,
      "top": -2,
    },
    Object {
      "baseSize": 3,
      "bottom": -Infinity,
      "bottomEdgeRow": -4,
      "distanceFromCenter": 1.118033988749895,
      "left": 3,
      "right": 5,
      "top": -3,
    },
    Object {
      "baseSize": 5,
      "bottom": -Infinity,
      "bottomEdgeRow": -4,
      "distanceFromCenter": 2.0615528128088303,
      "left": 2,
      "right": 6,
      "top": -4,
    },
  ],
  "closedVacancies": Array [],
  "leftEdgeVacancies": Array [],
  "rightEdgeVacancies": Array [],
  "topEdgeVacancies": Array [],
}
`;

exports[`VacanciesManager tests should return correct vacancies for sceneMap 8 1`] = `
Object {
  "bottomEdgeVacancies": Array [
    Object {
      "baseSize": 1,
      "bottom": -Infinity,
      "bottomEdgeRow": -4,
      "distanceFromCenter": 2.0615528128088303,
      "left": 2,
      "right": 2,
      "top": Infinity,
      "topEdgeRow": -1,
    },
  ],
  "closedVacancies": Array [],
  "leftEdgeVacancies": Array [],
  "rightEdgeVacancies": Array [],
  "topEdgeVacancies": Array [
    Object {
      "baseSize": 1,
      "bottom": -Infinity,
      "bottomEdgeRow": -4,
      "distanceFromCenter": 2.0615528128088303,
      "left": 2,
      "right": 2,
      "top": Infinity,
      "topEdgeRow": -1,
    },
  ],
}
`;
